File: binary_search/J.cpp
#include <iostream>

namespace aboba = std;

bool canPlace(long long x, long long w, long long h, long long n, long long a, long long b)
{
    return aboba::max((w / (a + 2 * x)) * (h / (b + 2 * x)), (h / (a + 2 * x)) * (w / (b + 2 * x))) >= n;
}

int main()
{
    long long n, a, b, w, h;

    aboba::cin >> n >> a >> b >> w >> h;

    long long left = 0, right = aboba::min(w, h);

    while (right - left > 1)
    {
        long long mid = (left + right) / 2;

        if (canPlace(mid, w, h, n, a, b))
            left = mid;
        else
            right = mid;
    }

    aboba::cout << left << aboba::endl;
}

File: string/J.py
def z_func(s):
    n = len(s)
    z = [0] * n
    left, right = 0, 0
    for i in range(1, n):
        k = max(0, min(z[i - left], right - i))
        while i + k < n and s[k] == s[i + k]:
            k += 1
        z[i] = k
        if i + z[i] > right:
            left, right = i, i + z[i]
    return z

n = int(input())
s = input()

s += s[::-1]

print(*z_func(s)[-1:n-1:-1])